components/build_image.py

```diff
@@ -0,0 +1,118 @@
+import argparse
+import datetime
+import logging
+import os
+import subprocess
+import sys
+import time
+import yaml
+
+
+def run(command, cwd=None, env=None, polling_interval=datetime.timedelta(seconds=1)):
+  """Run a subprocess.
+  Copied from kubeflow/test so it's easier to run locally.
+  Any subprocess output is emitted through the logging modules.
+  Returns:
+    output: A string containing the output.
+  """
+  logging.info("Running: %s \ncwd=%s", " ".join(command), cwd)
+
+  if not env:
+    env = os.environ
+  else:
+    keys = sorted(env.keys())
+
+    lines = []
+    for k in keys:
+      lines.append("{0}={1}".format(k, env[k]))
+    logging.info("Running: Environment:\n%s", "\n".join(lines))
+
+  process = subprocess.Popen(
+    command, cwd=cwd, env=env, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+
+  logging.info("Subprocess output:\n")
+  output = []
+  while process.poll() is None:
+    process.stdout.flush()
+    for line in iter(process.stdout.readline, ''):
+      output.append(line.strip())
+      logging.info(line.strip())
+
+    time.sleep(polling_interval.total_seconds())
+
+  process.stdout.flush()
+  for line in iter(process.stdout.readline, b''):
+    output.append(line.strip())
+    logging.info(line.strip())
+
+  if process.returncode != 0:
+    raise subprocess.CalledProcessError(process.returncode,
+                                        "cmd: {0} exited with code {1}".format(
+                                        " ".join(command), process.returncode), "\n".join(output))
+
+  return "\n".join(output)
+
+def build_tf_serving(args):
+  context_dir = "k8s-model-server/images"
+  version = args.tf_version if args.platform == "cpu" else args.tf_version + "gpu"
+
+  config_file = os.path.join(context_dir, "versions", version, "version-config.json")
+  with open(config_file) as f:
+    config = yaml.load(f)
+  if args.platform == "cpu":
+    command = [
+        "docker", "build", "--pull",
+        "--build-arg",  "TF_SERVING_PACKAGE={}".format(config['package']),
+        "-t", "{}/tensorflow-serving-{}:{}".format(args.registry, version, args.tag),
+        "-f", "Dockerfile.cpu", "."]
+  else:  # gpu
+    command = [
+        "docker", "build", "--pull",
+        "--build-arg", "BASE_IMAGE={}".format(config['base_image']),
```

Comment:
Would it be possible to automatically build-args from `version-config.json` instead of having to manually specify them here?

For example if `version-config.json` contains 

```
{
  "base_image": "nvidia/cuda:9.0-cudnn7-devel-ubuntu16.04",
  "tf_version": "1.7.0",
}
```

we can automatically add 2 `--build-args` for docker build:
```
--build-arg BASE_IMAGE=nvidia/cuda:9.0-cudnn7-devel-ubuntu16.04
--build-arg TF_VERSION=1.7.0
```